x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/EsqlSlowLogIT.java

```diff
@@ -0,0 +1,161 @@
+        for (int i = 0; i < 10; i++) {
+            testAllLevels(levels, coordinator, randomIntBetween(0, 10_000));
+        }
+    }
+
+    private static void testAllLevels(Map<Level, String> levels, DiscoveryNode coordinator, long timeoutMillis) throws InterruptedException,
+        ExecutionException {
+        for (Map.Entry<Level, String> logLevel : levels.entrySet()) {
+
+            client().execute(
+                ClusterUpdateSettingsAction.INSTANCE,
+                new ClusterUpdateSettingsRequest(TimeValue.THIRTY_SECONDS, TimeValue.THIRTY_SECONDS).persistentSettings(
+                    Settings.builder().put(logLevel.getValue(), timeoutMillis + "ms")
+                )
+            ).get();
+
+            EsqlQueryRequest request = EsqlQueryRequest.syncEsqlQueryRequest();
+            var querySource = "FROM index-* | EVAL ip = to_ip(host) | STATS s = COUNT(*) by ip | KEEP ip | LIMIT 100";
+            request.query(querySource);
+            request.pragmas(randomPragmas());
+            CountDownLatch latch = new CountDownLatch(1);
+            client(coordinator.getName()).execute(EsqlQueryAction.INSTANCE, request, ActionListener.running(() -> {
+                try {
+                    if (appender.lastEvent() == null) {
+                        if (timeoutMillis == 0) {
+                            fail("Expected a slow log with timeout set to zero");
+                        }
+                        return;
+                    }
+                    var msg = (ESLogMessage) appender.lastMessage();
+                    long took = Long.valueOf(msg.get("elasticsearch.slowlog.took"));
+                    long tookMillisExpected = took / 1_000_000;
+                    long tookMillis = Long.valueOf(msg.get("elasticsearch.slowlog.took_millis"));
+                    assertThat(took, greaterThan(0L));
+                    assertThat(tookMillis, greaterThanOrEqualTo(timeoutMillis));
+                    assertThat(tookMillis, is(tookMillisExpected));
+
+                    long planningTook = Long.valueOf(msg.get("elasticsearch.slowlog.planning.took"));
+                    long planningTookMillisExpected = planningTook / 1_000_000;
+                    long planningTookMillis = Long.valueOf(msg.get("elasticsearch.slowlog.planning.took_millis"));
+                    assertThat(planningTook, greaterThanOrEqualTo(0L));
+                    assertThat(planningTookMillis, greaterThanOrEqualTo(timeoutMillis));
+                    assertThat(planningTookMillis, is(planningTookMillisExpected));
+
+                    assertThat(took, greaterThan(planningTook));
+
+                    assertThat(msg.get("elasticsearch.slowlog.source"), is(querySource));
+                    assertThat(appender.getLastEventAndReset().getLevel(), equalTo(logLevel.getKey()));
+                } finally {
+                    latch.countDown();
+                }
+            }));
+            latch.await(30, TimeUnit.SECONDS);
```

Comment:
Are we failing the test if the countdown isn't 0? I think `await()` does nothing about it, right?
Not sure what will happen here exactly if the query fails.

I see other test doing an assert like:
```
assertEquals("All requests must respond, requests: " + requests, 0, latch.getCount());
```

Reply:
Please use `safeAwait(latch);` instead. It should handle correctly the cases when operation times out.

Reply:
>I see other test doing an assert like:

One more assert won't hurt, I'm adding it.